MAINT: Refactor of `numpy/core/_type_aliases.py`

mtsokol opened the pull request:
Hi @seberg @rgommers,

This is a successor for #24651 PR. It focuses on refactoring both C and Python code that defines types and aliases. `_type_aliases.py` imports `typeinfo` and moves types to `allTypes` and `sctypeDict` without the old logic, such as recalculating bit sizes etc.   

Here, `allTypes` contains concrete type names that are available in the main namespace and abstract types (also placed in the main namespace). `sctypeDict` contains concrete type names and additional aliases that are used by `np.dtype` (e.g. `np.dtype("long")` or `np.dtype("int")`). `sctypes` is exactly the same as before.

The only thing left is the refactor of `numpy/f2py` test suite, as it uses original `typeinfo` quite extensively. The rest of the test suites pass successfully on my machine.

seberg committed:
MAINT: Simplify typeinfo infrastructure by exposing dtype instances

This could probably be simplified more by using `np.dtypes` or
merging some of the information there, but this is a big first step
of removing logic duplication and just removing a lot of exposed
info that nobody needs.

mtsokol committed:
MAINT: Refactor allTypes, sctypeDict and sctypes build process

mtsokol committed:
MAINT: Refactor f2py/tests/test_array_from_pyobj.py

seberg commented on numpy/core/_type_aliases.py:
```diff
@@ -17,221 +17,87 @@
+    "object": "object_",
+    "bytes": "bytes_",
+    "int": "intp",
+    "long": "intp",
```
This looks wrong.  `long` has  nothing to do with `intp`, form the C side, you currently get the "current" name `int_`.  You could change that to long on the C-side though (since `int_` is not a nice name).  `int` is probably "default integer" which is `int_` which is `long`.

Once we change the default int, that will need a bit of shuffling around, although if we rename to `long` on the C-side, we can probably do all of that here in Python.

`intp` is its own thing, its included on the C-side dict I provided.

mtsokol committed:
Change int and long mapping

mtsokol commented:
Hi @seberg,

I've got two questions regarding remaining checks that failed:
1. Should I drop `np.float128` and `np.complex256` usages and replace them with `(c)longdouble`? The new implementation of `_type_aliases.py` doesn't build these entries for `allTypes` as it only imports names from `typeinfo`.
2. I'm wondering how to solve one last C dependency that causes tests to break. The `numpy/f2py/tests/test_array_from_pyobj.py` needs to know the C name of a given dtype [here](https://github.com/numpy/numpy/blob/4f84d719657eb455a35fcdf9e75b83eb1f97024a/numpy/f2py/tests/test_array_from_pyobj.py#L215) as it uses C name to import a dtype from `f2py` extension. Previously it worked because `typeinfo` was a dict with C names as keys and dtypes as values. 
As in the current implementation `typeinfo`s keys are already proper Python names I have no way to determine "C" name to use to call "wrap" `f2py` extension with.
Is there another way to get `C type name -> Python dtype` mapping to use for this test? (It looks like the mapping that we intentionally removed from `typeinfo` in this PR).

seberg commented:
Hmmm, I thought this was already being pushed forward/through, while I was out.

> Should I drop np.float128 and np.complex256 usages and replace them with (c)longdouble? The new implementation of _type_aliases.py doesn't build these entries for allTypes as it only imports names from typeinfo.

Just use `.itemsize` to define `float{...}` if the itemsize is 96 or 128, and the same for `clongdouble`?  I think we should at least add a warning on those names, but that is a distinct issue.

> I'm wondering how to solve one last C dependency that causes tests to break

I guess we could add the `NPY_INT`, etc. names to the dictionary, additionally, or a different dictionary.  In general, I don't mind the approach taken here to just have another dict, especially in tests where it isn't as bad to duplicate things.

seberg commented:
I would really like this to move forward, because I started looking into changing the definition for `intp`, which means changing a character code here or there probably.  So there are likely conflicts.

mtsokol commented:
@seberg Thank you for feedback! For now I will include these extended precision sized aliases based on `.itemsize`. 
I think that additional dict built on C side that defines C names mapping to dtypes (to be used in tests) solves this issue.

mtsokol committed:
Include c-names mapping

mtsokol commented:
Hi @seberg,

I pushed a new commit - here are my comments:
1. I added C's `NPY_*` names that are used in f2py tests to get `cname -> dtype` mapping that was missing.
2. I named it `c_names_dict`, is it fine, or would you prefer different name?
3. I added extended precision sized aliases as we discussed (`np.float128`, `np.complex256` etc.)
4. The PR is ready from my side. One question: I see that 6 CI stages are missing - all azure-pipelines - do you know why? (it would be better to see them passing as well here)

mattip commented:
I don't know why it is not showing up in the summary, but there is a [test failure](https://dev.azure.com/numpy/numpy/_build/results?buildId=32253&view=logs&j=94356751-2a09-5f35-07c1-87bf12815f8e&t=aed0f7cb-c656-5ca0-8fcd-90f7c267ae7c) on azure.

mtsokol commented:
> I don't know why it is not showing up in the summary, but there is a [test failure](https://dev.azure.com/numpy/numpy/_build/results?buildId=32253&view=logs&j=94356751-2a09-5f35-07c1-87bf12815f8e&t=aed0f7cb-c656-5ca0-8fcd-90f7c267ae7c) on azure.

Huh, the link you provided is from different PR (`core` to `_core` rename), not this one.
Maybe I can just retrigger.

mtsokol commented:
Ok, now azure stages are running.

ngoldbaum commented:
> I added extended precision sized aliases as we discussed

Are you OK making the bitsized longdouble aliases generate deprecation warnings in a followup?

ngoldbaum commented:
I was asking Mateusz since that was the only ask in the reviews that wasn't addressed. Thanks for the context though.

I looked over the C changes earlier and didn't see any issues. I'll give them another once-over before merging.

mtsokol commented:
Sure! I will deprecate these aliases in the next PR.

ngoldbaum reviewed (commented):
I looked at the C changes once more and didn't see any issues. Thanks for pushing this through @mtsokol @seberg!

mroeschke commented:
It appears this removed some previously public aliases in `np.dtype`. Is this expected?

```python
In [1]: import numpy as np

In [2]: np.dtype("unicode")
Out[2]: dtype('<U')

In [3]: np.__version__
Out[3]: '1.24.4'
```

```python
>>> import numpy as np
>>> np.dtype("unicode")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: data type 'unicode' not understood
>>> np.__version__
'2.0.0.dev0+git20231007.040ed2d'
```

ngoldbaum commented:
I think that was missed, can you open an issue?

mroeschke commented:
https://github.com/numpy/numpy/issues/24892
